src/librustc_trait_selection/traits/mod.rs

```diff
@@ -297,7 +297,9 @@ pub fn normalize_param_env_or_error<'tcx>(
     );
 
     let mut predicates: Vec<_> =
-        util::elaborate_predicates(tcx, unnormalized_env.caller_bounds.to_vec()).collect();
+        util::elaborate_predicates(tcx, unnormalized_env.caller_bounds.to_vec())
+            .map(|obligation| obligation.predicate)
+            .collect();
```

Comment:
Tempted to try and find every allocation in the trait system and rewrite it.

Reply:
Are you going to be doing it or should I try and do some of it now?

Reply:
If you want to play around with it, feel free to! I was mostly considering taking this over in case you don't have time for it.

Reply:
I won't have time to look at `elaborate_obligations`, which is the one that might have the biggest impact, but I can have a small PR shortly where I'm replacing a bunch of these arguments with `impl Iterator<Item = PredicateObligation>` so that it becomes clearer where we're `collect`ing multiple times.

Reply:
Looking at this in closer detail, I think I might be able to simplify this a bit.